flink-runtime/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java

```diff
@@ -188,6 +196,14 @@ private void expand() {
             }
         }
 
+        private <R> void repeatUntilConverged(Supplier<R> producer) {
+            R lastResult = producer.get();
+            R nextResult;
+            while (!lastResult.equals(nextResult = producer.get())) {
+                lastResult = nextResult;
+            }
+        }
+
         private List<Transformation<?>> getSinkTransformations(int sizeBefore) {
```

Comment:
Didn't we discuss offline that the current behavior of sublisting the transformations isn't safe?

Reply:
Yes that's why I create a copy now of the sublist (should be List.copyOf in newer JDK versions). If new transforms are added, we have a separate iteration with a separate copy that expands them.